README.md
# homebridge-sprinklerd

## Homebridge platform plugin for SprinklerD

## Installation prerequisites

1) Install [SprinklerD](https://github.com/sfeakes/SprinklerD)
2) Install [Homebridge](https://github.com/nfarina/homebridge) on any machine
3) Optional install [homebridge-config-ui-x](https://github.com/oznu/homebridge-config-ui-x)

## Installation

**Option 1: Install via Homebridge Config UI X:**

 1. Navigate to the Plugins page in in [homebridge-config-ui-x](https://github.com/oznu/homebridge-config-ui-x).
 2. Search for "sprinklerd" and install homebridge-aqualinkd.

**Option 2: Manually Install:**
 1. Install this on same machine as homebridge.

```
sudo npm install -g homebridge-sprinklerd
```

## Update

**Option 1: Update via Homebridge Config UI X:**

 1. Navigate to the Plugins page in [homebridge-config-ui-x](https://github.com/oznu/homebridge-config-ui-x).
 2. Click the Update button for the SprinklerD plugin.

**Option 2: Manually Update:**
```
sudo npm update -g homebridge-sprinklerd
```

## Configuration

**Option 1: Configuration via Homebridge Config UI X:**

 1. Navigate to the Plugins page in homebridge-config-ui-x.
 2. Click the Settings button for the SprinklerD plugin.

**Option 2: Manually Configuration:**

Edit homebridge config.json<br>
Example config
```
"platforms": [
        {
            "platform": "sprinklerd",
            "name": "SprinklerD",
            "server": "my-server-or-ip-running-sprinklerd",
            "port": "80",
            "mqtt": {
              "host": "my-mqtt-server",
              "port": 1883,
              "topic": "sprinklerd"
            }
       }
    ],
```

CHANGELOG.md
# Change Log

## v0.1.2 (2023-06-27)
### Featured Changes
* Fixed node.js v18 issue where ipv6 address would get resolved from localhost rather than ipv4 address. (Only issue if SprinklerD service is running on the same machine as Homebridge.)


lib/mqtt.js
var mqtt = require('mqtt');
var platform;
var client;
var config = {host: "", port: 0, credentials: false, channel: ""};

module.exports = {
  Mqtt: Mqtt
}

function Mqtt(aPlatform, host, port, channel, credentials) {
  platform = aPlatform;

  config = {host: host, port: port, credentials: credentials, channel: channel};
  if (typeof config.credentials === 'undefined' || typeof config.credentials.username === 'undefined' || config.credentials.username.length == 0) {
    config.credentials = false;
  }
  this.connect();
}

Mqtt.prototype.connect = function() {
    var connectOptions = {
    host: config.host,
    port: config.port
  };

  if (config.credentials)
  {
    connectOptions.username = config.credentials.username;
    connectOptions.password = config.credentials.password;
  }

  platform.forceLog("Connecting MQTT broker to "+config.host+":"+config.port+" with topic "+config.channel);

  client = mqtt.connect(connectOptions);

  client.on('connect', function() {
    platform.forceLog("Successfully connected to MQTT broker.");
    client.subscribe(config.channel+"/#");
  });

  client.on('close', function(error) {
    client.end(true, function() {
      this.error("Retrying in 5 seconds...");
      setTimeout(function() {
        platform.forceLog("Retrying connection to MQTT broker...");
        this.connect();
      }.bind(this), 5000);
    }.bind(this));

  }.bind(this));

  client.on('error', function(error) {
    platform.forceLog("ERROR connecting to MQTT broker.");
    client.end(true, function() {
      this.error(error);
    }.bind(this));
  }.bind(this));

  client.on('message', function (topic, buffer) {
      //platform.log("MQTT Message "+topic.toString() + " state "+buffer.toString());
      var accessory = platform.accessories.find(function(acc) {
        if (topic.substring(topic.lastIndexOf("/")+1) == "set" ) {
          return false;
        } else if (config.channel.length == topic.lastIndexOf("/")) {
          id = topic.substring(config.channel.length+1);
        } else {
          id = topic.substring(config.channel.length+1, topic.lastIndexOf("/"));
        }
        return acc.id == id;
      });
      
      if (!accessory) {
        return;
      }
      accessory.handleMQTTMessage(topic.toString(), buffer.toString(), function(characteristic, value) {
        if (typeof value !== 'undefined' && typeof characteristic !== 'undefined') {
          characteristic.setValue(value, null, "sprinklerd-MQTT");
        }
      });
  });
}

Mqtt.prototype.send = function(topic, message) {
  if (client)
  {
    platform.log("MQTT send topic:'"+config.channel+"/"+topic+"' message:''"+message+"'");
    client.publish(config.channel+"/"+topic, message);
  }
}

Mqtt.prototype.error = function(error) {
  var logMessage = "Could not connect to MQTT broker! (" + config.host + ":" + config.port + ")\n";

  if (config.credentials !== false) {
    logMessage += "Note: You're using a username and password to connect. Please verify your username and password too.\n";
  }

  if (error) {
    logMessage += error;
  }

  platform.forceLog(logMessage);
};

lib/utils.js
var inherits = require('util').inherits;

module.exports = {
    Utils: Utils
}

function Utils() {}


Utils.LogConnectionError = function(platform, message, err)
{
  var errorMessage = "There was a problem connecting to SprinklerD.";

  if (message) {
    errorMessage = message;
  }

  
  if (err && err.statusCode) {
    errorMessage += " (HTTP Status code " + err.statusCode + ")\n" + err.body;
  } else if (err && err.statusText) {
    errorMessage += " : `" + err.statusText + "`";
  } else if (err && err.message) {
    errorMessage += " : `" + err.message+ "`";
  } else if (err) {
    errorMessage += "\n - " + err;
  }

  platform.forceLog.error(errorMessage);
}

var TRUTHY_VALUES = 'y yes true'.split(/\s/);
Utils.toBool =  function(value) {
    value = value.toString();
    value = value.trim();
    value = value.toLowerCase();

    // Empty string is considered a falsy value
    if(!value.length) {
      return false;

    // Any number above zero is considered a truthy value
    } else if(!isNaN(Number(value))) {
      return value > 0;

    // Any value not marked as a truthy value is automatically falsy
    } else {
      return TRUTHY_VALUES.indexOf(value) >= 0;
    }
}


lib/sprinkler.js

// We do not need node-fetch, if you remove this it will default to NodeJS internal fetch.
// BUT node-fetch Gives way more meaningful errors than moment, so leaving it in.
var fetch = require("node-fetch"); 

// We need to set ipv4 first, latest node is ipv6 first. 
// can to this be resetting dns or custom http.agent
//const dns = require('node:dns');
// Use http ageng over above.
const http = require('node:http');



//var request = require("request");
var Constants = require('./constants.js');
var Utils = require('./utils.js').Utils;
// var extend = require('util')._extend;

module.exports = {
  Sprinkler : Sprinkler
}

var httpAgent = false;

function Sprinkler() {}

Sprinkler.initialize = function(useSSL, requestHeaders) {
  // We can force ipv4 first with below.
  //dns.setDefaultResultOrder('ipv4first');
  // Or we can use a custo, http.user agent.
  httpAgent = new http.Agent({
    family: 4
  });
};

Sprinkler.devices = async function(baseURL, completion, error) {

  if (httpAgent === false) {
    return;
  }
  
  var url = baseURL + "type=homebridge";
  var res;

  try {
    res = await fetch(url, {agent: httpAgent} );
  } catch (err) {
    if (typeof error !== 'undefined' && error !== false) {
      error("Error connecting to SprinklerD service", err);
    } else {
      console.log("Error connecting to SprinklerD service : "+ err);
    }
    return;
  }

  if (!res.ok) {
    if (typeof error !== 'undefined' && error !== false) {
      error("Error returned from SprinklerD service", res);
    } else {
      console.log(`Error returned from SprinklerD service: ${res.statusText}`);
    }
    return;
  } else {
    try {
      const json = await res.json();
      var devices = [];

      if (json.devices === undefined) {
        if (typeof completion !== 'undefined' && completion !== false) {
          completion(devices);
        }
        return;
      }

      for (var i = 0; i < json.devices.length; i++) {
        var device = json.devices[i];
        // Reset name on master valve.
        if (device.id == Constants.idMasterValve)
          device.name = json.title;

        devices.push(device);
      }

      if (typeof completion !== 'undefined' && completion !== false) {
        completion(devices);
      }
    } catch (err) {
      if (typeof error !== 'undefined' && error !== false) {
        error("Error understanding result from SprinklerD service", err);
      } else {
        console.log("Error understanding result from SprinklerD service : "+ err);
      }
      return;
    }
  } 
};
/*
Sprinkler.devices = function(baseURL, completion, error) {
  if (baseHttpRequest === false) {
    return;
  }

  var url = baseURL + "type=homebridge";

  baseHttpRequest.get({url : url}, function(err, response, json) {
    if (!err && response.statusCode == 200) {
      var devices = [];

      if (json.devices === undefined) {
        if (typeof completion !== 'undefined' && completion !== false) {
          completion(devices);
        }
        return;
      }

      for (var i = 0; i < json.devices.length; i++) {
        var device = json.devices[i];
        // Reset name on master valve.
        if (device.id == Constants.idMasterValve)
          device.name = json.title;

        devices.push(device);
      }

      if (typeof completion !== 'undefined' && completion !== false) {
        completion(devices);
      }
    } else {
      if (typeof error !== 'undefined' && error !== false) {
        error(response, err);
      }
    }
  });
};
*/

/*
// This is not used, just here incase we go back to mising HTTP with MQTT
Sprinkler.deviceStatus = function(accessory, completion, statustype, error) {
  if (baseHttpRequest === false) {
    return;
  }

  var url = accessory.platform.apiBaseURL + "type=homebridge";

  baseHttpRequest.get({url : url}, function(err, response, json) {

    if (!err && response.statusCode == 200 && json !== undefined) {
      if (json.devices === undefined) {
        // Handle error
        return;
      }
      for (var i = 0; i < json.devices.length; i++) {
        var device = json.devices[i];
        if (device.id == accessory.id) {
          if (statustype === Constants.statusDuration)
            completion(device.duration);
          else 
            completion(device.state == "on" ? 1 : 0);
        }
      }
    } else {
      Utils.LogConnectionError(this.platform, response, err);
      if (typeof error !== 'undefined' && error !== false) {
        error();
      }
    }
  }.bind(accessory));
};
*/
Sprinkler.updateDeviceStatus = function(accessory, value, completion, statustype = Constants.statusStatus) {
  if (accessory.platform.mqtt) {
    if (statustype === Constants.statusDuration)
      accessory.platform.mqtt.send(accessory.id + "/duration/set", value.toString());
    else
      accessory.platform.mqtt.send(accessory.id + "/set", value ? "1" : "0");
    if (typeof completion !== 'undefined' && completion !== false) {
      completion(true);
    }
    return;
  }
/*
  var url = accessory.platform.apiBaseURL + "type=option&option=" + this.id + "&state=" + value ? "on" : "off";
  if (accessory.type == Constants.sdDeviceZone && this.id != Constants.idCycleAllZones)
    url = accessory.platform.apiBaseURL + "type=zone&zone=" + this.zonenumber + "&state=" + value ? "on" : "off";

  for (var key in parameters) {
    url += "&" + encodeURI(key) + "=" + encodeURI(parameters[key]);
  }

  Sprinkler.updateWithURL(accessory, url, completion);
*/
};

/*
Sprinkler.updateWithURL = function(accessory, url, completion) {
  if (baseHttpRequest === false) {
    return;
  }

  baseHttpRequest.put({url : url, header : this.requestHeaders, json : true}, function(err, response) {
    var success = (typeof err === 'undefined' || !err);

    if (success) {
      this.platform.log(this.name + " sent command succesfully.");
    } else {
      Utils.LogConnectionError(this.platform, response, err);
    }

    if (typeof completion !== 'undefined' && completion !== false) {
      completion(success);
    }
  }.bind(accessory));
};
*/


lib/sprinklerd_accessory.js
// var request = require("request");
var Constants = require('./constants.js');
// var Helper = require('./helper.js').Helper;
// var Domoticz = require('./domoticz.js').Domoticz;
// var eDomoticzServices = require('./services.js').eDomoticzServices;
// module.exports = eDomoticzAccessory;
var Mqtt = require('./mqtt.js').Mqtt;
var Utils = require('./utils.js').Utils;

var Sprinkler = require('./sprinkler.js').Sprinkler;
module.exports = SprinklerdAccessory;

function SprinklerdAccessory(platform, platformAccessory, id, state, type, zonenumber, name, duration, uuid) {
  this.services = [];
  this.platform = platform;
  this.zonenumber = zonenumber;
  this.name = name;
  this.type = type;
  this.id = id;
  //state == "on" ? this.state = true : this.state = false;
  this.state = Utils.toBool(state);
  this.duration = duration;
  this.remainingduration = 0;

  var voidCallback = function() {};

  this.platformAccessory = platformAccessory;
  if (!this.platformAccessory) {
    this.platformAccessory = new platform.api.platformAccessory(this.name, uuid);
  }
  this.platformAccessory.reachable = true;
  this.publishServices();
}

SprinklerdAccessory.prototype = {
  identify : function(callback) { callback(); },
  publishServices : function() {
    var services = this.getServices();
    for (var i = 0; i < services.length; i++) {
      var service = services[i];

      var existingService = this.platformAccessory.services.find(function(eService) { return eService.UUID == service.UUID; });

      if (!existingService) {
        this.platformAccessory.addService(service, this.name);
      }
    }
  },
  getService : function(name) {
    var service = false;
    try {
      service = this.platformAccessory.getService(name);
    } catch (e) {
      service = false;
    }

    if (!service) {
      var targetService = new name();
      service = this.platformAccessory.services.find(function(existingService) { return existingService.UUID == targetService.UUID; });
    }

    return service;
  },
  getCharacteristic : function(service, name) {
    var characteristic = false;
    try {
      characteristic = service.getCharacteristic(name);
    } catch (e) {
      this.platform.forceLog("^ For: " + this.name + " " + name.AccessoryInformation);
      characteristic = false;
    }

    if (!characteristic) {
      var targetCharacteristic = new name();
      characteristic = service.characteristics.find(function(existingCharacteristic) { return existingCharacteristic.UUID == targetCharacteristic.UUID; });
    }

    return characteristic;
  },
  gracefullyAddCharacteristic : function(service, characteristicType) {
    var characteristic = this.getCharacteristic(service, characteristicType);
    if (characteristic) {
      return characteristic;
    }

    return service.addCharacteristic(new characteristicType());
  },
  setState : function(state, callback, context) {
    // console.log("sprinklerd_accessory.setState '"+this.id+"' "+state+" '"+context+"'");
    if (context && context == "sprinklerd-MQTT") {
      //this.state = state == 1 ? true : false;
      this.state = Utils.toBool(state);
      if (this.state === false)
        this.remainingduration = 0;
      callback();
      return;
    }

    Sprinkler.updateDeviceStatus(this, state, function(success) { callback(); }.bind(this));
    // this.platform.mqtt.send(this.id+"/set", state?"1":"0");
    // callback();
  },
  getState : function(callback) {
    // console.log("sprinklerd_accessory.getState '"+this.id+"' "+this.state);
    callback(null, this.state);
    /*
    Sprinkler.deviceStatus(this, Constants.statusState, function (value) {
        if (!this.state) {
            callback(null, value);
        }
        this.state = value;
    }.bind(this));
    */
  },
  setDuration : function(value, callback, context) {
    // console.log("sprinklerd_accessory.setValue '"+this.id+"' "+value+" '"+context+"'");

    if (context && context == "sprinklerd-MQTT") {
      this.duration = value;
      callback();
      return;
    }

    // Set duration for every valve if we are "cycleallzones" button
    if (this.id == Constants.idCycleAllZones) {
      for (i = 1; i < this.zonenumber; i++) {
        this.platform.mqtt.send("zone" + i + "/duration/set", value.toString());
      }
      callback();
    } else { // det duration for single valve
      // this.platform.mqtt.send(this.id+"/duration/set", value.toString());
      Sprinkler.updateDeviceStatus(this, value, function(success) { callback(); }.bind(this), Constants.statusDuration);
    }
  },
  getDuration : function(callback) {
    // console.log("sprinklerd_accessory.getValue '"+this.id+"' "+this.duration);
    callback(null, this.duration);
    return;

    /*
    Sprinkler.deviceStatus(this, Constants.statusDuration, function (value) {
        if (!this.duration) {
            callback(null, value);
        }
        this.duration = value;
    }.bind(this));
    */
  },

  setRemainingDuration : function(valveService, value, callback, context) {
    // console.log("sprinklerd_accessory.setRemainingDuration '"+this.id+"' "+value);

    this.remainingduration = value;
    this.getCharacteristic(valveService, Characteristic.RemainingDuration).updateValue(value);
    callback();

  },
  getRemainingDuration : function(callback) {
    // console.log("sprinklerd_accessory.getRemainingDuration '"+this.id+"' "+this.remainingduration);
    callback(null, this.remainingduration);
    return;
  },
  getServices : function() {
    this.services = [];
    var informationService = this.getService(Service.AccessoryInformation);
    if (!informationService) {
      informationService = new Service.AccessoryInformation();
    }
    informationService.setCharacteristic(Characteristic.Manufacturer, "sprinklerd")
        .setCharacteristic(Characteristic.Model, this.type)
        .setCharacteristic(Characteristic.SerialNumber, "Sprinklerd " + this.name);
    this.services.push(informationService);

    if (this.type === Constants.sdDeviceSwitch) {
      service = this.getService(Service.Switch);
      if (!service) {
        service = new Service.Switch(this.name);
      }
      this.getCharacteristic(service, Characteristic.On).on('set', this.setState.bind(this)).on('get', this.getState.bind(this));

    } else if (this.type === Constants.sdDeviceZone) {
      service = this.getService(Service.Valve);
      if (!service) {
        service = new Service.Valve(this.name);
      }
      service.setCharacteristic(Characteristic.ValveType, Characteristic.ValveType.IRRIGATION);
      // this.getCharacteristic(service, Characteristic.On).on('set', this.setState.bind(this)).on('get', this.getState.bind(this));
      this.getCharacteristic(service, Characteristic.Active).on('set', this.setState.bind(this)).on('get', this.getState.bind(this));
      this.getCharacteristic(service, Characteristic.InUse).on('set', this.setState.bind(this)).on('get', this.getState.bind(this));
      this.getCharacteristic(service, Characteristic.RemainingDuration)
          .on('get', this.getRemainingDuration.bind(this))
          .on('set', this.setRemainingDuration.bind(this, service));
      if (this.id != Constants.idMasterValve) {
        this.getCharacteristic(service, Characteristic.SetDuration).on('set', this.setDuration.bind(this)).on('get', this.getDuration.bind(this));
      } else {
        // This has no effect at present, but might in future
        /*
        this.getCharacteristic(service, Characteristic.SetDuration).setProps({
          stepValue: 60,
          minValue: 0,
          maxValue: 3600
        });
        */
      }
    }

    this.services.push(service);

    return this.services;
  },

  handleMQTTMessage : function(topic, message, callback) {
    this.platform.log("MQTT received for '%s'. Topic:'%s' Message:'%s'", this.id, topic, message);

    var value = parseInt(message) == 1 ? 1 : 0;
    var service = false;
    //var value = Utils.toBool(state);

    if (this.type === Constants.sdDeviceSwitch) {
      service = this.getService(Service.Switch);
      var characteristic = this.getCharacteristic(service, Characteristic.On);
    } else if (this.type === Constants.sdDeviceZone) {
      service = this.getService(Service.Valve);
      if (topic.substring(topic.lastIndexOf("/") + 1) == "duration") {
        var characteristic = this.getCharacteristic(service, Characteristic.SetDuration);
        value = parseInt(message);
      } else if (topic.substring(topic.lastIndexOf("/") + 1) == "remainingduration") {
        var characteristic = this.getCharacteristic(service, Characteristic.RemainingDuration);
        value = parseInt(message);
      } else {
        var characteristic = this.getCharacteristic(service, Characteristic.Active);
        callback(characteristic, value);
        var characteristic = this.getCharacteristic(service, Characteristic.InUse);
      }
    }

    callback(characteristic, value);
  }
}


index.js

// Example ~/.homebridge/config.json content:
//
// {
//  "bridge": {
//         "name": "Homebridge",
//         "username": "CC:21:3E:E4:DE:33", // << Randomize this...
//         "port": 51826,
//         "pin": "031-45-154", // << Change pin
//      },
//
//  "platforms": [{
//         "platform": "sprinklerd",
//         "name": "sprinklerd",
//         "server": "127.0.0.1", // servername/ip running SprinklerD
//         "port": "80",
//         "mqtt": {
 //              "host": "mqtt-server.name",  // servername/ip running MQTT
//               "port": 1883,
//               "topic": "sprinklerd"
//         }
//    }],
//  "accessories":[]
// }
//



var Sprinkler = require('./lib/sprinkler.js').Sprinkler;
var Mqtt = require('./lib/mqtt.js').Mqtt;
var Utils = require('./lib/utils.js').Utils;
var SprinklerdAccessory = require('./lib/sprinklerd_accessory.js');
var Constants = require('./lib/constants.js');

const util = require('util');

module.exports = function(homebridge) {
  Service = homebridge.hap.Service;
  Characteristic = homebridge.hap.Characteristic;
  Types = homebridge.hapLegacyTypes;
  UUID = homebridge.hap.uuid;

  homebridge.registerPlatform("homebridge-sprinklerd", "sprinklerd", SprinklerdPlatform, true);
};

function SprinklerdPlatform(log, config, api) {
  this.isSynchronizingAccessories = false;
  this.accessories = [];
  this.forceLog = log;
  this.log = function() {
    if (typeof process.env.DEBUG !== 'undefined') {
      log(util.format.apply(this, arguments));
    }
  };

  this.config = config;
  try {
    this.server = config.server;
  } catch (e) {
    return;
  }
  this.authorizationToken = false;
  if (this.server.indexOf(":") > -1 && this.server.indexOf("@") > -1) {
    var tmparr = this.server.split("@");
    this.authorizationToken = Helper.Base64.encode(tmparr[0]);
    this.server = tmparr[1];
  }
  this.port = config.port;
  this.api = api;
  this.apiBaseURL = "http://" + this.server + ":" + this.port + "/?";
  this.mqtt = false;
  this.ssl = false;

  var requestHeaders = {};
  if (this.authorizationToken) {
    requestHeaders['Authorization'] = 'Basic ' + this.authorizationToken;
  }
  Sprinkler.initialize(this.ssl, requestHeaders);

  if (this.api) {
    this.api.once("didFinishLaunching", function() {
      var syncDevices = function() {
        this.synchronizeAccessories();
        setTimeout(syncDevices.bind(this), 600000); // Sync devices every 10 minutes
        //setTimeout(syncDevices.bind(this), 6000); // Sync devices every 1 minutes
      }.bind(this);
      syncDevices();

      if (config.mqtt) {
        setupMqttConnection(this);
      }
    }.bind(this));
  }
}

SprinklerdPlatform.prototype = {
  synchronizeAccessories : function() {

    if (this.isSynchronizingAccessories) {
      return;
    }

    this.isSynchronizingAccessories = true;
    // var excludedDevices = (typeof this.config.excludedDevices !== 'undefined') ? this.config.excludedDevices : [];

    Sprinkler.devices(this.apiBaseURL,
                      function(devices) {
                        var removedAccessories = [], exclude = !1;

                        for (var i = 0; i < devices.length; i++) {
                          var device = devices[i];
                         
                          var existingAccessory = this.accessories.find(function(existingAccessory) { return existingAccessory.id == device.id; });
                          
                          if (existingAccessory) {
                            if (device.type != existingAccessory.type) {
                              this.log("Device " + existingAccessory.name + " has changed it's type. Recreating...");
                              removedAccessories.push(existingAccessory);
                              try {
                                this.api.unregisterPlatformAccessories("homebridge-sprinklerd", "sprinklerd", [ existingAccessory.platformAccessory ]);
                              } catch (e) {
                                this.forceLog("Could not unregister platform accessory! (" + removedAccessory.name + ")" + e);
                              }
                            } else {
                              // Since we have everything, may as well update if needed.
                              if ( (device.state=="on"?true:false) != existingAccessory.state ) {
                                this.log("Background update to '"+existingAccessory.id+"' state old="+existingAccessory.state+" new="+(device.state=="on"?true:false))
                                existingAccessory.state = (device.state==1?true:false);
                              }
                              if (device.duration != existingAccessory.duration) {
                                this.log("Background update to '"+existingAccessory.id+"' duration old="+existingAccessory.duration+" new="+device.duration);
                                existingAccessory.duration = device.duration;
                              }
                              if (existingAccessory.id == Constants.idCycleAllZones && existingAccessory.zonenumber != device.zone) {
                                this.log("Background update to '"+existingAccessory.id+"' zone# old="+existingAccessory.zonenumber+" new="+device.zone);
                                existingAccessory.zonenumber = device.zone;
                              }
                              
                              continue;
                            }
                          }
                          // Generate a new accessory
                          var uuid = UUID.generate(device.name);

                          var accessory = new SprinklerdAccessory(this, false, device.id, device.state, device.type, device.zone, device.name, device.duration, uuid);
                          this.accessories.push(accessory);
                          this.forceLog("Registering new platform accessory! (" + accessory.name + " | " + uuid + ")");

                          try {
                            this.api.registerPlatformAccessories("homebridge-sprinklerd", "sprinklerd", [ accessory.platformAccessory ]);
                          } catch (e) {
                            this.forceLog("Could not register platform accessory! (" + accessory.name + ")" + e);
                          }
                          // accessory.platformAccessory.context = {device: device, uuid: uuid, eve: this.eve};
                          accessory.platformAccessory.context = {device : device, uuid : uuid};
                        }

                        for (var i = 0; i < this.accessories.length; i++) {
                          var removedAccessory = this.accessories[i];
                          var existingDevice = devices.find(function(existingDevice) { return existingDevice.idx == removedAccessory.idx; });

                          if (!existingDevice) {
                            removedAccessories.push(removedAccessory);
                            try {
                              this.api.unregisterPlatformAccessories("homebridge-sprinklerd", "sprinklerd", [ removedAccessory.platformAccessory ]);
                            } catch (e) {
                              this.forceLog("Could not unregister platform accessory! (" + removedAccessory.name + ")" + e);
                            }
                          }
                        }

                        for (var i = 0; i < removedAccessories.length; i++) {
                          var removedAccessory = removedAccessories[i];
                          var index = this.accessories.indexOf(removedAccessory);
                          this.accessories.splice(index, 1);
                        }

                        this.isSynchronizingAccessories = false;
                      }.bind(this),
                      function(response, err) {
                        Utils.LogConnectionError(this, response, err);
                        this.isSynchronizingAccessories = false;
                      }.bind(this));
  },
  configureAccessory : function(platformAccessory) {
    if (!platformAccessory.context || !platformAccessory.context.device) {
      // Remove this invalid device from the cache.
      try {
        this.api.unregisterPlatformAccessories("homebridge-sprinklerd", "sprinklerd", [ platformAccessory ]);
      } catch (e) {
        this.forceLog("Could not unregister cached platform accessory!" + e);
      }
      return;
    }

    var device = platformAccessory.context.device;
    var uuid = platformAccessory.context.uuid;

    //console.log("Loading platform accessory! (" + device.name + " | " + uuid + ")");
    this.forceLog("Loading platform accessory [" + device.name + "]");

    // Generate the already cached accessory again
    var accessory = new SprinklerdAccessory(this, platformAccessory, device.id, device.state, device.type, device.zone, device.name, device.duration, uuid);
    this.accessories.push(accessory);
  }
};

function setupMqttConnection(platform) {
  var connectionInformation = {
    host : (typeof platform.config.mqtt.host !== 'undefined' ? platform.config.mqtt.host : '127.0.0.1'),
    port : (typeof platform.config.mqtt.port !== 'undefined' ? platform.config.mqtt.port : 1883),
    topic : (typeof platform.config.mqtt.topic !== 'undefined' ? platform.config.mqtt.topic : 'sprinklerd'),
    username : (typeof platform.config.mqtt.username !== 'undefined' ? platform.config.mqtt.username : ''),
    password : (typeof platform.config.mqtt.password !== 'undefined' ? platform.config.mqtt.password : ''),
  };

  platform.forceLog("MQTT Connect info : " + platform.config.mqtt.host + ":" + platform.config.mqtt.port + " | topic : " + platform.config.mqtt.topic);

  var mqttError = function() {
    platform.forceLog(
        "There was an error while getting the MQTT Hardware Device from Sprinklerd.\nPlease verify that you have started sprinklerd enabeled MQTT and configured zones.");
  };

  platform.mqtt = new Mqtt(platform, connectionInformation.host, connectionInformation.port, connectionInformation.topic,
                           {username : connectionInformation.username, password : connectionInformation.password});
}
